### ISSUE TITLE
CA2000 ConcurrentDictionary AddOrUpdate (false positive)

### ISSUE THREAD

**Issue by k1ll3r8e** [dotnet/roslyn-analyzers#3082]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.8 (Latest)

#### Diagnostic ID

Example: [CA2000](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2000?view=vs-2019)

#### Repro steps

1. Using .Net Framework 4.6
2. Adding a disposable type to a ConcurrentDictionary

```c#
    private readonly ConcurrentDictionary<uint, Channel> _chs = new ConcurrentDictionary<uint, Channel>();
    
    public IChannel GetChannel(int timeout)
    {
        if (Connected)
        {
            lock (syncThreadID)
            {
                if (_threadID > C_MAX)
                {
                    _threadID = 1;
                }
                uint id = _threadID++;
                while (_chs.ContainsKey(id))
                {
                    id = _threadID++;
                }
                return _chs.AddOrUpdate(id, new Channel(this, id, timeout), (k, v) => v);
            }
        }
        return null;
    }
```

#### Expected behavior

```c#
    return _chs.AddOrUpdate(id, new Channel(this, id, timeout), (k, v) => v);
```

The Channel is created, added to the ConcurrentDictionary and will be returned to the caller. So the caller has to call Dispose on it's own. The Dispose implementation in the "Channel" removes it from ConcurrentDictionary. Also, disconnecting/closing/disposing the Socket, where GetChannel(int timeout) is implemented, will raise an event which cause all "Channel"s to Dispose them self. So, in any case , the CA2000 is a false positive report in this situation.

#### Actual behavior

```c#
    return _chs.AddOrUpdate(id, new Channel(this, id, timeout), (k, v) => v);
```

A CA2000 error is emitted.

**Comment by mavasani** [comment]
https://dotnet.myget.org/feed/roslyn-analyzers/package/nuget/Microsoft.CodeAnalysis.FxCopAnalyzers/2.9.9-beta1.19610.1+906398e7 should have the fix.
